### ISSUE TITLE
Infix operators in the text format?

### ISSUE THREAD

**Issue by sunfishcode** [CraneStation/cranelift#513]
Currently, cranelift IR is always printed with one instruction per line, eg.:
```
function %foo(i32, i32, i32) -> i32 {
ebb0(v0: i32, v1: i32, v2:i32):
    v3 = imul v0, v1
    v4 = iadd v3, v2
    return v4
}
```

What if we introduced some simple syntax sugar for instructions with only one use? It'd be in addition to the existing syntax. We could then (optionally) print that same code like this:
```
function %foo(i32, i32, i32) -> i32 {
ebb0(v0: i32, v1: i32, v2:i32):
    return v0 * v1 + v2
}
```
That would be much easier to read in many cases, which is of potential interest to cranelift developers, but also to cranelift users looking to understand how cranelift is compiling their code.

This also might make it even more interesting to switch to `//` comments (#471).

There's some ambiguity with syntax like `v0 + 1`, but I think we can resolve it by saying that we always use the `_imm` instruction when possible rather than emitting an iconst

And there's the question if value numbers for the intermediate values. My rough idea is that they'd just always use the next available value number.

There are other issues to consider too, such as printing srclocs and instruction encodings. But I think we could find reasonable ways to make these work. The main question is, is this idea worth pursuing?

**Comment by Hyperion101010** [comment]
Hi there sunfishcode can i work on this issue, i am new to contributing for open source.
I found compiler design interesting in our class and decided to contribute for same.
I have contributed for mozilla ,some patches in past few months, so i know the procedure of adding patches :)

**Comment by Hyperion101010** [comment]
If i am not wrong, web assembly will be used by high level language programmers(web developers) that use js in their day-to-day work. So the code must be easy to read. Moreover if we are thinking about games,etc developers they code in C++ which is again not assembly.

**Comment by sunfishcode** [comment]
Hello! This is a good issue to work on, though to be clear, this is about the text format for Cranelift IR, not for WebAssembly itself, so it won't be visible to most developers unless they use compiler-explorer-type tools to look at the compiler IR as its compiling their code.

One of the big tasks here will be to implement infix expression parsing. My high-level idea here is that we can use an operator precedence parser. Do you think that sounds like a good approach?

**Comment by Hyperion101010** [comment]
Let's discuss on irc ! 
I am very excited to start this issue

**Comment by Hyperion101010** [comment]
hi sunfish what tool are we going to use here ? I can practice them to get working on this issue !

**Comment by sunfishcode** [comment]
My rough idea here is that a hand-written [operator precedence parser](https://en.wikipedia.org/wiki/Operator-precedence_parser) would work well, since the rest of the parser is also hand-written. However, I'm open to other suggestions here.

**Comment by phase** [comment]
I don't think this is something than an IR should add. This adds needless complexity that compiler writers don't _need_ when working with IR. It seems odd that we would want to obfuscate the registers when that's exactly what a compiler writer would want to see.

**Comment by Hyperion101010** [comment]
As far as i know it is just a debug feature for programmer .Using this he can understand wasm code more clearly like high level code .

**Comment by sunfishcode** [comment]
@phase When I see something like "add v3, v4" in an IR dump before register allocation, most often my next question is "what defined r3 and/or r4?" and next most often my next question is "what else is using v3 or v4?" The infix notation proposed here would make it easier to answer both those questions in many common cases.

If we make this optional so that people don't have to see it when they don't want to, and if we disable it during and after register allocation, it seems like it could make the IR much more readable in many common cases.

Ultimately, I can't say for sure if this will be great, but it has potential :-).

**Comment by bnjbvr** [comment]
> @phase When I see something like "add v3, v4" in an IR dump before register allocation, most often my next question is "what defined r3 and/or r4?" and next most often my next question is "what else is using v3 or v4?" The infix notation proposed here would make it easier to answer both those questions in many common cases.

This made me think that an alternative / a nice addition to solve this particular problem would be to implement the [language server protocol](https://microsoft.github.io/language-server-protocol/overview) for CLIF. Then, we get "jump to definition" and "see all uses" for free in all the IDEs that have LSP addons (Visual Code, Sublime Text, Vim, etc.). Totally over-engineering to solve such a use case, but a fun small, self-contained project to write, and this could reuse a lot of code from the CLIF parsing infrastructure.
